src/main/java/org/opensearch/agent/tools/DataDistributionTool.java

```diff
@@ -0,0 +1,1304 @@
+     */
+    private String getPPLQueryWithTimeRange(String query, String startTime, String endTime, String timeField) {
+        if (Strings.isEmpty(timeField)) {
+            return query;
+        }
+
+        String formattedStartTime = formatTimeForPPL(startTime);
+        String formattedEndTime = formatTimeForPPL(endTime);
+        String timePredicate = String
+            .format(Locale.ROOT, "`%s` >= '%s' AND `%s` <= '%s'", timeField, formattedStartTime, timeField, formattedEndTime);
+
+        if (Strings.isEmpty(query)) {
+            return "WHERE " + timePredicate;
+        }
+
+        String[] commands = query.split("\\|");
+        for (int i = 0; i < commands.length; i++) {
+            String cmd = commands[i].trim();
+            if (cmd.regionMatches(true, 0, "where", 0, 5)) {
+                commands[i] = cmd + " AND " + timePredicate;
+                return String.join(" | ", Arrays.stream(commands).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList()));
+            }
+        }
+
+        // No WHERE found: insert after first command
+        List<String> commandList = new ArrayList<>();
+        commandList.add(commands[0].trim());
+        commandList.add("WHERE " + timePredicate);
+        for (int i = 1; i < commands.length; i++) {
+            String cmd = commands[i].trim();
+            if (!cmd.isEmpty()) {
+                commandList.add(cmd);
+            }
+        }
+        return String.join(" | ", commandList);
+    }
+
+    /**
+     * Builds PPL query string for data retrieval within specified time range
+     *
+     * @param index Index name
+     * @param timeField Time field name
+     * @param startTime Start time for query
+     * @param endTime End time for query
+     * @param size Maximum number of documents
+     * @param customPpl Custom PPL statement (optional)
+     * @return Formatted PPL query string
+     */
+    private String buildPPLQuery(String index, String timeField, String startTime, String endTime, int size, String customPpl) {
+        String baseQuery;
+
+        if (!Strings.isEmpty(customPpl)) {
+            baseQuery = getPPLQueryWithTimeRange(customPpl, startTime, endTime, timeField);
+        } else {
+            baseQuery = getPPLQueryWithTimeRange(String.format(Locale.ROOT, "source=%s", index), startTime, endTime, timeField);
+        }
+
+        return baseQuery + String.format(Locale.ROOT, " | head %d", size);
+    }
+
+    /**
+     * Executes PPL query and parses the result using provided result parser
+     *
+     * @param <T> The parsed result type
+     * @param ppl PPL query string to execute
+     * @param resultParser Function to parse PPL result into desired format
+     * @param listener Action listener for handling parsed results or failures
+     */
+    private <T> void executePPLAndParseResult(String ppl, Function<Map<String, Object>, T> resultParser, ActionListener<T> listener) {
```

Comment:
We should already have this function in `LogPatternAnalysisTool`. Can we reuse this? Any difference?